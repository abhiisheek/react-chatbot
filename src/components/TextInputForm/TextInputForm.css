@import '../_variables.css';
@import '../_colors.css';
@import '../_defaults.css';

.wrapper {
  border-top: 2px solid var(--border-color);
  padding: 10px;
  min-height: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: white;
}

.textInput {
  width: calc(100% - var(--send-btn-size));
  border: none;
  background-color: inherit;
}

.textInput:focus {
  outline: none;
}

.sendBtn {
  padding: 0px;
  border: none;
  background: transparent;
}

.sendBtn:focus {
  outline: none;
}

.upArrow::before {
  @apply --send-btn-icon;
}

.sendBtn:hover {
  cursor: pointer;
}

.disabled, 
.disabled:hover {
  @apply --disabled-btn;
}