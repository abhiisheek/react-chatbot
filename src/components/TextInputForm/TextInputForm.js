import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';

import styles from './TextInputForm.css';

/** Component which takes the users input and provides an mechanism for handling it */
class TextInputForm extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      value: ''
    };
  }

  handleOnClick = () => {
    this.props.onUserInput(this.state.value);
    this.setState({
      value: ''
    });
  };

  handleOnChange = event => this.setState({ value: event.target.value });

  handleOnKeyPress = event => {
    if (event.key === 'Enter') {
      this.handleOnClick();
    }
  };

  render() {
    return (
      <div className={styles.wrapper}>
        <input
          className={styles.textInput}
          value={this.state.value}
          type="text"
          placeholder="Type here . . ."
          onChange={this.handleOnChange}
          onKeyPress={this.handleOnKeyPress}
        />
        <button
          className={classNames(styles.sendBtn, {
            [`${styles.disabled}`]: this.state.value === ''
          })}
          onClick={this.handleOnClick}
          disabled={this.state.value === ''}
        >
          <div className={styles.upArrow} />
        </button>
      </div>
    );
  }
}

TextInputForm.displayName = 'TextInputForm';

TextInputForm.propTypes = {
  /** Handler for handling user input */
  onUserInput: PropTypes.func.isRequired
};

export default TextInputForm;
