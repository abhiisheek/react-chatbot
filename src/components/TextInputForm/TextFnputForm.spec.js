import React from 'react';
import renderer from 'react-test-renderer';

import { shallow } from 'enzyme';

import TextInputForm from './TextInputForm';

const handleUserInput = jest.fn();

const createElement = () =>
  shallow(<TextInputForm onUserInput={handleUserInput} />);

describe('TextInputForm', () => {
  it('Should have displayName set on the component', () => {
    expect(TextInputForm.displayName).toBeDefined();
  });

  it('Snapshot test', () => {
    const snapshot = renderer
      .create(<TextInputForm onUserInput={handleUserInput} />)
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('onChangeHandler should be called whenever user gives inputs', () => {
    const wrapper = createElement();
    const component = wrapper.instance();
    const spy = jest.spyOn(component, 'handleOnChange');
    const event = { target: { value: 'user input here' } };
    // Note: Needed both, for spying
    // Refer : https://github.com/airbnb/enzyme/issues/586
    component.forceUpdate();
    wrapper.update();
    wrapper.find('input').simulate('change', event);
    expect(spy).toHaveBeenCalled();
    expect(spy).toBeCalledWith(event);
  });

  it('User input should be considered and updated in the state', () => {
    const wrapper = createElement();
    wrapper
      .find('input')
      .simulate('change', { target: { value: 'user input here' } });
    expect(wrapper.state('value')).toBe('user input here');
  });

  it('onKeyPressHandler should be called whenever user gives inputs', () => {
    const wrapper = createElement();
    const component = wrapper.instance();
    const spy = jest.spyOn(component, 'handleOnKeyPress');
    const event = { key: 'A' };
    component.forceUpdate();
    wrapper.update();
    wrapper.find('input').simulate('keypress', event);
    expect(spy).toHaveBeenCalled();
    expect(spy).toBeCalledWith(event);
  });

  it('handleOnClick should be called whenever user presses ENTER key', () => {
    const wrapper = createElement();
    const component = wrapper.instance();
    const spy = jest.spyOn(component, 'handleOnClick');
    const event = { key: 'Enter' };
    component.forceUpdate();
    wrapper.update();
    wrapper.find('input').simulate('keypress', event);
    expect(spy).toHaveBeenCalled();
  });

  it('Should call handleUserInput with user input when onClickHandler is clicked/called', () => {
    const wrapper = createElement();
    wrapper.setState({
      value: 'User input'
    });
    wrapper.instance().handleOnClick();
    expect(handleUserInput).toHaveBeenCalled();
    expect(handleUserInput).toBeCalledWith('User input');
  });

  it('Should clear input field value once user input is submitted', () => {
    const wrapper = createElement();
    wrapper.setState({
      value: 'User input'
    });
    wrapper.instance().handleOnClick();
    expect(wrapper.find('input').text()).toBe('');
    expect(wrapper.state('value')).toBe('');
  });

  it('Send button should be disabled when the user input is empty', () => {
    const wrapper = createElement();
    expect(wrapper.find('.sendBtn').hasClass('disabled')).toBeTruthy();
    expect(wrapper.find('.sendBtn').prop('disabled')).toBeTruthy();
  });

  it('handleOnClick should be called whenever user presses send button', () => {
    const wrapper = createElement();
    const component = wrapper.instance();
    const spy = jest.spyOn(component, 'handleOnClick');
    component.forceUpdate();
    wrapper.update();
    wrapper.find('.sendBtn').simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
