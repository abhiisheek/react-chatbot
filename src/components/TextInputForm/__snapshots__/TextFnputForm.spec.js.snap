// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputForm Snapshot test 1`] = `
<div
  className="wrapper"
>
  <input
    className="textInput"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Type here . . ."
    type="text"
    value=""
  />
  <button
    className="sendBtn disabled"
    disabled={true}
    onClick={[Function]}
  >
    <div
      className="upArrow"
    />
  </button>
</div>
`;
